internal/getproviders/didyoumean_test.go

```diff
@@ -98,19 +134,45 @@ func TestMissingProviderSuggestion(t *testing.T) {
 			ctx,
 			want,
 			source,
+			Requirements{
+				want: MustParseVersionConstraints(">= 1.0.0"),
+			},
 		)
 		if got != want {
 			t.Errorf("wrong result\ngot:  %s\nwant: %s", got, want)
 		}
+
+		// Checking the case where we have a default namespace
+		// but a provider in the requirements that can fulfill it
+		wantDiffNamespace := addrs.Provider{
+			Hostname:  svchost.Hostname("example.com"),
+			Namespace: "nuhuh",
+			Type:      "foo",
+		}
+		gotDiffNamespace := MissingProviderSuggestion(
+			ctx,
+			addrs.Provider{
+				Hostname:  svchost.Hostname("example.com"),
+				Namespace: "hashicorp",
+				Type:      "foo",
+			},
+			source,
+			Requirements{
+				wantDiffNamespace: MustParseVersionConstraints(">= 1.0.0"),
+			},
+		)
+		if gotDiffNamespace != wantDiffNamespace {
+			t.Errorf("wrong result\ngot:  %s\nwant: %s", gotDiffNamespace, wantDiffNamespace)
+		}
```

Comment:
I don't quite understand the higher-level goal of this test (and the corresponding change from using `addr.IsDefault()` to checking the namespace directly), but I might be missing something.

The situation I think we're trying to address in this PR arises when there is no `required_providers` entry for a provider, resulting in an implied default provider from a resource type or provider block. However I believe that the situation in this test case is only possible if we have two providers with explicit source addresses like `example.com/nuhuh/foo` and `example.com/hashicorp/foo`. As far as I'm aware, there's no way to get into the situation under test here without an explicit `required_providers` source address for these providers.

If possible, I think it would be preferable to keep leaning on the `IsDefault` semantics, so that the two lines of logic line up. We'd only make "did you mean…?" suggestions when the provider source address is at least possibly implied, and if the user specifies a source address, we assume they know what they're doing.

Reply:
ah! ok I will adjust that behavior accordingly then!

Reply:
Latest commit goes back to `IsDefault`. To re-articulate for myself, it would be unusual for someone to have a registry failure for their own registry for an implied provider, because in order for it to be their own registry provider definition, they would have had to have a `required_provider` block, so to get the [pre-new commit] situation, they'd have to define a required providers block specifically with the default namespace [incorrectly]?